src/main/java/io/netty/incubator/codec/http3/QpackDecoder.java

```diff
@@ -16,94 +16,383 @@
+     */
+    public void insertWithNameReference(QuicStreamChannel qpackDecoderStream,
+                                        boolean staticTableRef, int nameIdx, CharSequence value) throws QpackException {
+        final QpackHeaderField entryForName;
+        if (staticTableRef) {
+            entryForName = QpackStaticTable.getField(nameIdx);
         } else {
-            throw new UnsupportedOperationException("dynamic table is not implemented");
+            entryForName = dynamicTable.getEntryRelativeEncoderInstructions(nameIdx);
         }
+        dynamicTable.add(new QpackHeaderField(entryForName.name, value));
+        sendInsertCountIncrementIfRequired(qpackDecoderStream);
+    }
+
+    /**
+     * Inserts a header field with a literal name corresponding to the
+     * <a href="https://quicwg.org/base-drafts/draft-ietf-quic-qpack.html#name-insert-with-literal-name">
+     *     encoder instruction.</a>
+     *
+     * @param qpackDecoderStream {@link QuicStreamChannel} for the QPACK decoder stream.
+     * @param name of the field.
+     * @param value of the field.
+     * @throws QpackException if the insertion fails.
+     */
+    public void insertLiteral(QuicStreamChannel qpackDecoderStream, CharSequence name, CharSequence value)
+            throws QpackException {
+        dynamicTable.add(new QpackHeaderField(name, value));
+        sendInsertCountIncrementIfRequired(qpackDecoderStream);
+    }
+
+    /**
+     * Duplicates a previous entry corresponding to the
+     * <a href="https://quicwg.org/base-drafts/draft-ietf-quic-qpack.html#name-insert-with-literal-name">
+     *     encoder instruction.</a>
+     *
+     * @param qpackDecoderStream {@link QuicStreamChannel} for the QPACK decoder stream.
+     * @param index which is duplicated.
+     * @throws QpackException if duplication fails.
+     */
+    public void duplicate(QuicStreamChannel qpackDecoderStream, int index)
+            throws QpackException {
+        dynamicTable.add(dynamicTable.getEntryRelativeEncoderInstructions(index));
+        sendInsertCountIncrementIfRequired(qpackDecoderStream);
+    }
+
+    /**
+     * Callback when a bi-directional stream is
+     * <a href="https://quicwg.org/base-drafts/draft-ietf-quic-qpack.html#name-abandonment-of-a-stream"> abandoned</a>
+     *
+     * @param qpackDecoderStream {@link QuicStreamChannel} for the QPACK decoder stream.
+     * @param streamId which is abandoned.
+     */
+    public void streamAbandoned(QuicStreamChannel qpackDecoderStream, long streamId) {
+        if (maxTableCapacity == 0) {
+            return;
+        }
+        // https://quicwg.org/base-drafts/draft-ietf-quic-qpack.html#name-stream-cancellation
+        //   0   1   2   3   4   5   6   7
+        // +---+---+---+---+---+---+---+---+
+        // | 0 | 1 |     Stream ID (6+)    |
+        // +---+---+-----------------------+
+        final ByteBuf cancel = qpackDecoderStream.alloc().buffer();
```

Comment:
nit: we can do better in terms of initial size of the buffer here... 8 bytes should be enough for all values